frontend/src/lib/keybind.ts

```diff
@@ -0,0 +1,246 @@
+   *
+   * Will run the given callback if all requirements are met from an item in the
+   * previously given {@link KeybindCallback} array.
+   *
+   * @param e - The `keydown` event of `KeyboardEvent` type
+   * @param useKeyboardShortcuts - A boolean of the user's `useKeyboardShortcuts` preference
+   */
+  onKeyDown(e: KeyboardEvent, useKeyboardShortcuts: boolean) {
+    if (!useKeyboardShortcuts || e.repeat) return;
+
+    const enum_key = this.keyEnumFromString(e.key);
+
+    if (!(enum_key && this.bindingEntries.includes(enum_key))) return;
+
+    // Conditionals to be able to later compare to a potentially undefined binding.shift/.al/.ctrl
+    const shift = e.shiftKey ? true : undefined;
+    const ctrl = e.ctrlKey ? true : undefined;
+    const alt = e.altKey ? true : undefined;
+
+    const binding = this.bindings.find((binding) => binding.key === enum_key);
+    if (binding && binding.alt == alt && binding.shift == shift && binding.ctrl == ctrl) {
+      e.preventDefault();
+      binding.callback(e);
+    }
+  }
+}
+
+/*
+Search callbacks
+*/
+
+/**
+ * Used to indicate what direction {@link navigateResults} should go (up/down)
+ */
+type Direction = 1 | -1;
+
+/**
+ * Utilized to bring the next/previous result into focus
+ *
+ * @remarks
+ * Utilizes the element that is currently in focus and the given
+ * direction to determine where and how to traverse the array of
+ * results. Moves a single result per call.
+ *
+ * @param direction - How to maneuver through the result list (up/down), indicated by {@link Direction}
+ */
+const navigateResults = (direction: Direction) => {
+  const currentActive = document.activeElement;
+
+  const results = [...document.getElementsByClassName('result')] as HTMLElement[];
+  let currentResultIndex: number | null = null;
+  let newIndex = 0;
+
+  results.forEach((el, index) => {
+    if (el.contains(currentActive)) currentResultIndex = index;
+  });
+
+  if (currentResultIndex != null) {
+    newIndex = currentResultIndex + direction;
+    if (0 > newIndex || newIndex >= results.length) {
+      newIndex = currentResultIndex;
+    }
+  }
+
+  const nextResult = results[newIndex];
+
+  // Get the anchor element and focus it
+  (nextResult.getElementsByClassName('result-main-link')[0] as HTMLElement)?.focus();
+};
+
+/**
+ * Wrapper for {@link navigateResults} that will focus on the next result
+ */
+const focusNextResult = () => navigateResults(1);
+
+/**
+ * Wrapper for {@link navigateResults} that will focus on the previous result
+ */
+const focusPrevResult = () => navigateResults(-1);
+
+/**
+ * Focus the first result (if any)
+ */
+const focusMainResult = () => {
+  const results = document.getElementsByClassName('result-main-link');
+  if (results.length > 0) {
+    (results[0] as HTMLElement).focus();
+  }
+};
+
+/**
+ * Focus the search bar
+ */
+const focusSearchBar = () => {
+  (document.getElementById('searchbar') as HTMLInputElement)?.select();
```

Comment:
should take `Refs` as a parameter (with `SearchBar` added to the type) and use this instead of `getElemtById`